xdsl/dialects/builtin.py

```diff
@@ -397,21 +397,26 @@ def verify_value(self, value: int):
                 f"values in the range [{min_value}, {max_value})"
             )
 
-    def normalized_value(self, value: IntAttr) -> IntAttr | None:
+    def normalized_value(
+        self, value: IntAttr, *, truncate_bits: bool = False
+    ) -> IntAttr | None:
         """
         Signless values can represent integers from both the signed and unsigned ranges
         for a given bitwidth.
         We choose to normalize values that are not in the intersection of the two ranges
         to the signed version (meaning ambiguous values will always be negative).
         For example, the bitpattern of all ones will always be represented as `-1` at
         runtime.
-        If the input value is outside of the valid range, return `None`.
+        If the input value is outside of the valid range, return `None` if `truncate_bits` is false,
+        otherwise returns a value in range by truncating the bits of the input.
         """
         min_value, max_value = self.value_range()
         if not (min_value <= value.data < max_value):
-            return None
+            if not truncate_bits:
+                return None
+            value = IntAttr(value.data & (2**self.bitwidth - 1))
 
-        if self.signedness.data == Signedness.SIGNLESS:
+        if self.signedness.data != Signedness.UNSIGNED:
```

Comment:
why this change?

Reply:
Doing the bitwise and seems to wrap everything into the range [0,2**bitwidth), so for the signed case we need to normalise it back to the correct range. This seemed like the easiest way to do this but it could be refactored